### ISSUE TITLE
MIOS binding re-applies transformations for each item

### ISSUE THREAD

**Issue by beowulfe** [openhab/openhab#3081]
If I have the same MIOS service bound to multiple items, with transforms defined for each item, each transform receives as input the output of the previous transformation.

For example, I'm trying to use an RGB Bulb exposed over service/RGBController1/Color. The color is actually defined as 5 components - R, G, B, Warm Light Level, Cool Light Level. To map this to OpenHAB, I've created three items:

Color LightColor "Light Color" {mios="unit:house,device:15/service/RGBController1/Color,command:JS(miosRgbCommand.js),in:JS(miosRgbIn.js)"}
Dimmer WarmLight "Warm Light Level"{mios="unit:house,device:15/service/RGBController1/Color,command:JS(miosWarmLightCommand.js),in:JS(miosWarmLightIn.js)"}
Dimmer CoolLight "Cool Light Level" {mios="unit:house,device:15/service/RGBController1/Color,command:JS(miosCoolLightCommand.js),in:JS(miosCoolLightIn.js)"}

The input that miosWarmLightIn.js receives is the output of miosRgbIn.js, and miosCoolLightIn.js receives the output of miosWarmLightIn.js.

My expectation is that each transform would receive the value sent by MIOS. However, I wanted to check with @mrguessed before submitting a PR to resolve this - it's possible my expectation is based on an incorrect understanding of how other people are using this.

**Comment by mrguessed** [comment]
So you're probably hitting a couple of different issues here.

When you bind something like:

    unit:house,device:15/service/RGBController1/Color

It's saying that you want the `Color` UPnP _StateVariable_, from the MiOS's `RGBController1` serviceId _Alias_ you've defined, to be placed into that Item.

In this case, and looking at [the defn](http://forum.micasaverde.com/index.php?topic=32613.0), it'll have a value something like:
```
   #FF0000
```

The binding will take that value, run any `in` Transformation you've defined upon it, and then [poke](https://en.wikipedia.org/wiki/PEEK_and_POKE) it into the associated Item.

If you have multiple Items all bound to the same _StateVariable_, then they'll all receive the same value as input.

ie. `\#FF0000`, 

Each then applies it's own Transform to it, in your case they'll all need to filter and/or transform the bits they're interested in.

In this case, you've got `in:JS(miosRgbIn.js)` so I'm assuming that `miosRgbIn.js` has all the logic required to turn _\#RRGGBB????_ into the _HH,SS,BB_ that openHAB will need for input to the `Color` item it's bound to.

Something similar happens in reverse with the `command:` stuff.  They'll be presented with the native openHAB value, in String format, and the Transform mechanism will need to turn it into the `\#FF0000` style Color value that the MiOS RGB Plugin requires as input.

... and this is where it gets tricky, because something like the `WarmLight` Dimmer Item only has a partial (0-100) type value of what's going on.  It's been run through an `in`put filter you've defined and turned into a 0-100 value appropriate for a Dimmer Item (slider).  

If the user activates that Dimmer, then it's `command` filter will need to turn that into the [Color]RGB?? value needed by MiOS, and that might be visually jarring as the light suddenly changes from whatever RGB value it's on, to warm-white (at some intensity)

But that might be what you intended.  I'm guessing that the MiOS RGB Plugin has a few more State Variables that can be bound to (like separate R, G and B values, along with 

... then I read the code, and apparently it doesn't.  Oh well.



BTW: I have a Fibaro RGBW in my system, and I just bound the Items directly to what MiOS provided for me.

It ends up with a bunch of Items (generally via the _MiOS Item Generator_), but I tend to set-n-forget most of them.  I don't use the openHAB Color picker, since I find it a little finicky.  Anyhow, the generic set of mappings looks like:

```
Number   MasterClosetFibaroLightId "ID [%d]" {mios="unit:house,device:431/id"}
String   MasterClosetFibaroLightDeviceStatus "Master Closet Fibaro Light Device Status [%s]" {mios="unit:house,device:431/status"}
String   MasterClosetFibaroLightConfigured "Master Closet Fibaro Light Configured [%s]" {mios="unit:house,device:431/service/HaDevice1/Configured"}
DateTime MasterClosetFibaroLightLastUpdate "Master Closet Fibaro Light Last Update [%1$ta, %1$tm/%1$te %1$tR]" <calendar> {mios="unit:house,device:431/service/HaDevice1/LastUpdate"}
DateTime MasterClosetFibaroLightFirstConfigured "Master Closet Fibaro Light First Configured [%1$ta, %1$tm/%1$te %1$tR]" <calendar> {mios="unit:house,device:431/service/HaDevice1/FirstConfigured"}
Switch   MasterClosetFibaroLightStatus "Master Closet Fibaro Light" (GSwitch,GManagedEmergency,GManagedSwitch) {mios="unit:house,device:431/service/SwitchPower1/Status"}
Dimmer   MasterClosetFibaroLightLoadLevelStatus "Master Closet Fibaro Light Load Level Status [%d %%]" (GDimmer) {mios="unit:house,device:431/service/Dimming1/LoadLevelStatus"}
Dimmer   MasterClosetFibaroLightLoadLevelTarget "Master Closet Fibaro Light Load Level Target [%d %%]" {mios="unit:house,device:431/service/Dimming1/LoadLevelTarget"}
Number   MasterClosetFibaroLightWatts "Master Closet Fibaro Light Watts [%.1f W]" {mios="unit:house,device:431/service/EnergyMetering1/Watts"}
String   MasterClosetFibaroLightActualUsage "Master Closet Fibaro Light FIXME ActualUsage [%s]" {mios="unit:house,device:431/service/EnergyMetering1/ActualUsage"}
Number   MasterClosetFibaroLightKWH "Master Closet Fibaro Light kWh [%.1f kWh]" {mios="unit:house,device:431/service/EnergyMetering1/KWH"}
DateTime MasterClosetFibaroLightKWHReading "Master Closet Fibaro Light FIXME KWHReading [%1$ta, %1$tm/%1$te %1$tR]" <calendar> {mios="unit:house,device:431/service/EnergyMetering1/KWHReading"}
Number   MasterClosetDimmerId "ID [%d]" {mios="unit:house,device:432/id"}
String   MasterClosetDimmerDeviceStatus "Master Closet Dimmer Device Status [%s]" {mios="unit:house,device:432/status"}
Switch   MasterClosetDimmerStatus "Master Closet Dimmer" (GSwitch) {mios="unit:house,device:432/service/SwitchPower1/Status"}
Dimmer   MasterClosetDimmerLoadLevelStatus "Master Closet Dimmer Load Level Status [%d %%]" (GDimmer) {mios="unit:house,device:432/service/Dimming1/LoadLevelStatus"}
Dimmer   MasterClosetDimmerLoadLevelTarget "Master Closet Dimmer Load Level Target [%d %%]" {mios="unit:house,device:432/service/Dimming1/LoadLevelTarget"}
Number   MasterClosetRedLightId "ID [%d]" {mios="unit:house,device:433/id"}
String   MasterClosetRedLightDeviceStatus "Master Closet Red Light Device Status [%s]" {mios="unit:house,device:433/status"}
Switch   MasterClosetRedLightStatus "Master Closet Red Light" (GSwitch) {mios="unit:house,device:433/service/SwitchPower1/Status"}
Dimmer   MasterClosetRedLightLoadLevelStatus "Master Closet Red Light Load Level Status [%d %%]" (GDimmer) {mios="unit:house,device:433/service/Dimming1/LoadLevelStatus"}
Dimmer   MasterClosetRedLightLoadLevelTarget "Master Closet Red Light Load Level Target [%d %%]" {mios="unit:house,device:433/service/Dimming1/LoadLevelTarget"}
Number   MasterClosetBlueLightId "ID [%d]" {mios="unit:house,device:435/id"}
String   MasterClosetBlueLightDeviceStatus "Master Closet Blue Light Device Status [%s]" {mios="unit:house,device:435/status"}
Switch   MasterClosetBlueLightStatus "Master Closet Blue Light" (GSwitch) {mios="unit:house,device:435/service/SwitchPower1/Status"}
Dimmer   MasterClosetBlueLightLoadLevelStatus "Master Closet Blue Light Load Level Status [%d %%]" (GDimmer) {mios="unit:house,device:435/service/Dimming1/LoadLevelStatus"}
Dimmer   MasterClosetBlueLightLoadLevelTarget "Master Closet Blue Light Load Level Target [%d %%]" {mios="unit:house,device:435/service/Dimming1/LoadLevelTarget"}
Number   MasterClosetGreenLightId "ID [%d]" {mios="unit:house,device:434/id"}
String   MasterClosetGreenLightDeviceStatus "Master Closet Green Light Device Status [%s]" {mios="unit:house,device:434/status"}
Switch   MasterClosetGreenLightStatus "Master Closet Green Light" (GSwitch) {mios="unit:house,device:434/service/SwitchPower1/Status"}
Dimmer   MasterClosetGreenLightLoadLevelStatus "Master Closet Green Light Load Level Status [%d %%]" (GDimmer) {mios="unit:house,device:434/service/Dimming1/LoadLevelStatus"}
Dimmer   MasterClosetGreenLightLoadLevelTarget "Master Closet Green Light Load Level Target [%d %%]" {mios="unit:house,device:434/service/Dimming1/LoadLevelTarget"}
Number   MasterClosetWhiteLightId "ID [%d]" {mios="unit:house,device:436/id"}
String   MasterClosetWhiteLightDeviceStatus "Master Closet White Light Device Status [%s]" {mios="unit:house,device:436/status"}
Switch   MasterClosetWhiteLightStatus "Master Closet White Light" (GSwitch) {mios="unit:house,device:436/service/SwitchPower1/Status"}
Dimmer   MasterClosetWhiteLightLoadLevelStatus "Master Closet White Light Load Level Status [%d %%]" (GDimmer) {mios="unit:house,device:436/service/Dimming1/LoadLevelStatus"}
Dimmer   MasterClosetWhiteLightLoadLevelTarget "Master Closet White Light Load Level Target [%d %%]" {mios="unit:house,device:436/service/Dimming1/LoadLevelTarget"}
```

You may want to have some of this discussion on the forum, since it'll straighten stuff out fairly quickly.

**Comment by beowulfe** [comment]
The Aeotec bulb actually handles the transition between an RGB color and warm/cold nicely. In my command transforms, I simply clear out the values not being set. It's not perfect, because I lose the previous state for those unrelated items, but I can live with that. (Mios isn't creating separate UPNP devices for each RGB channels from the Aeotec like it seems to with your Fibaro)

The issue with the in: transforms stacking looks like it stems from https://github.com/openhab/openhab/blob/30a48547334b775764f284d44cc2a9b590d5a81e/bundles/binding/org.openhab.binding.mios/src/main/java/org/openhab/binding/mios/internal/MiosBinding.java#L455 where the value is being reused in subsequent iterations of the item loop. If I'm right, and that's not the behavior you intended, I can submit a PR.

**Comment by mrguessed** [comment]
If there are _different_ Items, then they should each run their _specific_ Transform of the original source-value from MiOS.

The intent was to use the same _Transformed_ MiOS source-value for all Bindings attached to the _same_ Item.

It's quite possible I got that logic incorrect, but I'd need to try it out first (I was guessing a bunch of the openHAB stuff, since it wasn't clear)

**Comment by mrguessed** [comment]
@beowulfe,
The application of the transform multiple times (chained) was in fact buggy, it shouldn't have done that.  I saw that in action while enhancing the Lock handling logic, and this is now fixed in the PR associated with #3317.

Many thanks for pointing it out!
